keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -118,78 +119,79 @@ Currently, kubeadm vendors in pieces of Go code from the components to be able t
 
-Kubeadm shall extend its current Kustomize support to include patching of core Kubernetes component configs. Currently, this will include only kubelet’s component config. However, in the future this might extend to the API Server, Scheduler and Controller Manager as well.  
-Most notably, this feature is not going to be applied to kube-proxy’s config by kubeadm as this is an addon component config and the responsibility for this is going to be with the corresponding addon operator.
+In addition to that, `kubeadm upgrade plan` is now also going to support a new flag `--component-config-upgrades` that is going to supply a file with one or more YAML documents in it that overwrite component configs that are too old and cannot be loaded from the cluster.  
+Only the configs that need upgrading would be loaded from the file supplied with `--component-config-upgrades`. Component configs that don't need version upgrades or kubeadm configuration types will be ignored.  
+The loaded configs, that were supplied with `--component-config-upgrades` would be taken into account when printing the additional component config information.
 
-Kustomize patches will be applied on both user supplied and kubeadm generated core component configs, just before they are stored in the local files from which the config will be consumed by the components. In the case of user supplied core component configs, neither the Kustomize patches, nor the result after applying them is going to be stored on the cluster. The config maps are going to contain only the vanilla user supplied component configs.
+### "kubeadm upgrade apply" changes
 
-If a user has provided Kustomize patches for a config version, that is different from the current one, kubeadm is going to bail out with an error, prompting the user to either delete the old patches, or to upgrade them to the new version. This is best implemented using a kubeadm pre-flight check.
+As supported component configs are expected to increase in number in the future and some of them might require manual user migration prior to upgrade, a handy and secure method for supplying new configs is necessary.
 
-UX wise, this feature is going to be reusing the existing kubeadm options that were introduced with the implementation of the kubeadm advanced configurations KEP. This allows for kustomizations to be applied to component configs on a per-node basis during init, join and upgrade just before the config is supposed to be stored on the local node file.
+For that matter `kubeadm upgrade apply` also gains the same `--component-config-upgrades` flag and with the same semantics as in `kubeadm upgrade plan`.  
+In addition to that, the newly used configs, from the file supplied with `--component-config-upgrades`, would be stored in the cluster as user specified configs during the upgrade process.
 
-### Stop using component config internal types
+### Introduce the "kubeadm alpha component-configs print-upgradeable" command
```

Comment:
an alpha command would mean that it has to graduate into a non-alpha command at some point.

should we use `kubeadm config` as the graduation target? i think having a separate `kubeadm component-configs` is not needed. so why not use e.g. `kubeadm alpha config print-upgradeable`?

Reply:
That's an interesting question to which I have no definitive answer ATM. I am too inclined to think that the place after graduation should be somewhere under `kubeadm config`.
However, we should make sure that the command refers to component configs specifically.
The whole point of having `print-upgradeable` is because we cannot automatically migrate some component configs. However, we can do this for kubeadm configs (`kubeadm config migrate`).
Hence, having `kubeadm config migrate` and `kubeadm config print-upgradeable`, or even `kubeadm config print upgradeable`, can be confusing to users.

Reply:
i'm proposing `kubeadm alpha config print ...` for the new alpha command.

Reply:
For a start, we are only concerned with the `upgrade apply` workflow for now.

1. User runs `kubeadm upgrade plan` and is prompted by the table at the end of it that a component config (or several ones) need manual upgrade.
2. (optional, but useful) User uses `kubeadm alpha component-configs print-upgradeable` to get the contents of upgradeable component configs. This is done for a reference.
3. User runs `kubeadm alpha component-configs print-defaults` to generate defaulted component configs for the current setup in a local file. Then manually edits them in the file to apply his/her own changes.
Alternatively, users may supply the configs in their entirety here in a local file, without using `kubeadm alpha component-configs print-defaults`.
4. (optional) Users re-run `kubeadm upgrade plan --config=<updated-configs-file>` to verify that the table now says that everything is OK.
5. Users continue on the normal upgrade path by using `kubeadm upgrade apply --config=<updated-configs-file>`
The new configs are automatically stored on the cluster upon successful completion of `upgrade apply`.

The rest of the upgrade process stays the same.

Reply:
i think the workflow is fine. there isn't much that can be done other than allowing them to fetch the defaults / upgradeable and manually edit, for now.